### DISCUSSION TITLE
Installation Issue - Avatar missing with error

### DISCUSSION THREAD

**Discussion by tinjaw** [outline/outline#2447] [General]
![CleanShot 2021-08-13 at 08 13 04](https://user-images.githubusercontent.com/110943/129362750-68dd4787-315a-48e0-98fc-e7573a2f6995.jpeg)
I used CapRover and installed outline. When I login I am missing my avatar and get an error. PLS see screenshot for detains.

**Comment by tommoor** [comment]
Please provide server logs, I'm guessing your storage (S3) is not setup correctly

**Comment by EacAyf** [comment]
Avatars are saved in the public folder, try adding read-only access rule to the public/* prefix.

**Comment by tinjaw** [comment]
I did an install without knowing anything about outline. And it looks like that is the problem. I was unaware that I needed to set up any integration with S3. Based on your replies above I peaked at the CapRover App Configs tab and see that there are boxes to enter config info for S3. I was never prompted/instructed to provide that info.

As soon as I get a moment free, I will create a bucket and add the info and report back. Thanks.

**Reply by tinjaw** [reply]
BTW, does it have to be specifically Amazon S3, or can I use any compatible service?

**Reply by EacAyf** [reply]
I personally run Minio in a docker container, had some issues setting it up in combination with Nginx as a reverse proxy. But works great now

**Reply by MetaBliss** [reply]
> I personally run Minio in a docker container, had some issues setting it up in combination with Nginx as a reverse proxy. But works great now

@EacAyf Can you share the issues you faced using Nginx Proxy Mgr with Outline to get the images uploading through MinIO? I'm facing a similar problem, Outline is fully setup but I have issues pasting images in my wiki or updating the profile or team pictures.

**Reply by EacAyf** [reply]
Tackling this issue was a long time ago, so had to dig in my memories. Any way, my issue came down to two things, both due to me and not outline's code. One was the 'abstract' S3 permissions and the second was CORS, as using a Nginx-proxy in combination with different subdomains. 

To be honest, and this may be different for you, I opted to make my bucket public allowing me to use my browser and Nginx logging to pick away at my CORS issue first. Once that was solved and I knew nothing was blocking the images being uploaded or displayed I switched into securing my minio bucket. As my outline wasn't filled with anything and it was only internally accessible this was pragmatic route. Still a lot of troubleshooting ensued. 

### Nginx
In the end I came to this solution, for both subdomains in my server bracket I explicitly push `add_header X-Frame-Options "SAMEORIGIN" always;` 
while in the location bracket I make sure it never reaches my subdomain with `proxy_hide_header X-Frame-Options`.

Writing this, and thinking about it, I might want to change this is the future... 

### Minio 
in my .env file of outline regarding Minio:
```
AWS_ACCESS_KEY_ID=<KEY>
AWS_SECRET_ACCESS_KEY=<SECRET>
AWS_REGION=eu-central-1
AWS_S3_UPLOAD_BUCKET_URL=https://<Minio.FQDN>
AWS_S3_UPLOAD_BUCKET_NAME=outline
AWS_S3_UPLOAD_MAX_SIZE=26214400
AWS_S3_FORCE_PATH_STYLE=true
AWS_S3_ACL=private

```
In Minio I created a User, assigned it to a group, created a policy and also assigned this to the group. I'm no expert so, if you see any security hardening tips, please let me know. This is the policy I created (for my outline bucket):
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws:s3:::outline/*"
            ]
        }
    ]
}

```
I then created the bucket, called outline, with a custom access policy:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetBucketLocation"
            ],
            "Resource": [
                "arn:aws:s3:::outline"
            ]
        },
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::outline"
            ],
            "Condition": {
                "StringEquals": {
                    "s3:prefix": [
                        "public"
                    ]
                }
            }
        },
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::outline/public*"
            ]
        }
    ]
}

```
So the way I understand it, again I may be wrong, is that the access key gives outline full rights on the outline bucket. Public users are only allowed to view images in the public folder of the bucket, the rest is off limits. 

p.s. step one was of course checking for silly mistakes, like 'mini' where it should have been 'minio'...
